### PR TITLE
splitting assert statements and further use from assert horizons.session.Session

### PR DESCRIPTION
**Pull request by MarkusHackspacher**
I take a look to the assert statements, I split it for a better diagnose.

And check where is use the assert horizons.session.Session with git grep horizons.session.Session.
In the file world/**init**.py have I activated and got this If I click at random map, for a new game.
Or at [Travis CI](https://travis-ci.org/MarkusHackspacher/unknown-horizons/builds/166106795):
python2 run_tests.py tests.gui.ingame.test_windows:test_settings_dialog_crash2 -a gui
python2 run_tests.py tests.gui.menu.test_multiplayer:test_create_game -a gui
python2 run_tests.py tests.gui.menu.test_singleplayer:test_start_map -a gui

```
 File "/unknown-horizons/horizons/gui/modules/singleplayermenu.py", line 425, in _update_map_preview
    world = load_raw_world(map_file)
 File "/unknown-horizons/horizons/world/__init__.py", line 741, in load_raw_world
    world = World(session=None)
 File "/unknown-horizons/horizons/world/__init__.py", line 86, in __init__
    assert isinstance(session, horizons.session.Session)
AttributeError: 'module' object has no attribute 'session'
```

### CODE REVIEW

File: horizons/world/__init__.py [hunk 1]
```diff
@@ -75,16 +75,15 @@ class World(BuildingOwner, WorldObject):
...
 	log = logging.getLogger("world")
 
 	def __init__(self, session):
-		"""
-		@param session: instance of session the world belongs to.
-		"""
 		self.inited = False
-		if False:
-			assert isinstance(session, horizons.session.Session)
+		# assert isinstance(session, horizons.session.Session)
```
Comment (MarkusHackspacher) [inline horizons/world/__init__.py]: If it active the tests not work:
https://travis-ci.org/MarkusHackspacher/unknown-horizons/builds/166106795

File: horizons/component/tradepostcomponent.py [hunk 2]
```diff
@@ -171,8 +173,8 @@ def sell(self, res, amount, price, player_id):
 			remnant = self.get_inventory().alter(res, -amount)
 			assert remnant == 0
 			self.trade_history.append((Scheduler().cur_tick, player_id, res, -amount, price))
-			self.sell_history[ Scheduler().cur_tick ] = (res, amount, price)
-			self.total_income += amount*price
```
Comment (MasterofJOKers) [inline horizons/component/tradepostcomponent.py]: This has nothing to do with the rest, doesn't it? Please keep you commits focused on the task they shall do.

File: horizons/gui/ingamegui.py [hunk 3]
```diff
@@ -454,7 +458,7 @@ def on_key_press(self, action, evt):
 		keyval = evt.getKey().getValue()
 
 		if action == _Actions.ESCAPE:
-			return self.on_escape()
```
Comment (MasterofJOKers) [inline horizons/gui/ingamegui.py]: Same here.

File: horizons/world/player.py [hunk 4]
```diff
@@ -69,7 +69,8 @@ def initialize(self, inventory):
 
 	def __init(self, name, color, clientid, difficulty_level, max_tier_notification, settlerlevel=0):
 		assert isinstance(color, Color)
-		assert isinstance(name, basestring) and name
+		assert isinstance(name, basestring)
+		assert name
```
Comment (MasterofJOKers) [inline horizons/world/player.py]: If you're adding the name, why don't you add `clientid`, `difficulty_level` and `max_tier_notification`?

File: horizons/world/player.py [hunk 5]
```diff
@@ -38,7 +38,16 @@
...
+	"""Class representing a player
+
+	@type session: horizons.session.Session
+	@param session: Session instance
+	@param worldid: player's worldid
+	@param name: user-chosen name
+	@param color: color of player (as Color)
+	@param clientid: id of client
+	@param inventory: {res: value} that are put in the players inventory
+	"""
```
Comment (MasterofJOKers) [inline horizons/world/player.py]: Why move the comment from `__init__()`?

File: horizons/world/__init__.py (hunk 1 above)
Comment (MasterofJOKers) [inline horizons/world/__init__.py]: Can you find out, why? Are you going to fix this?

File: horizons/world/player.py [hunk 6]
```diff
@@ -47,16 +56,7 @@ class Player(ComponentHolder, WorldObject):
...
-		@param session: Session instance
-		@param worldid: player's worldid
-		@param name: user-chosen name
-		@param color: color of player (as Color)
-		@param clientid: id of client
-		@param inventory: {res: value} that are put in the players inventory
-		"""
-		if False:
-			assert isinstance(session, horizons.session.Session)
+		assert isinstance(session, horizons.session.Session)
```
Comment (MasterofJOKers) [inline horizons/world/player.py]: So this assert is save now, while it was commented out earlier. What changed?

File: horizons/gui/ingamegui.py (hunk 3 above)
Comment (MarkusHackspacher) [inline horizons/gui/ingamegui.py]: That change make my IDE by there self, according to the .editorconfig settings.

Comment (squiddy) [inline horizons/gui/ingamegui.py]: I can confirm, I've had stripped whitespace between commits as well. Perhaps we should push a single commit removing trailing whitespace and be done with it.

Comment (squiddy) [inline horizons/gui/ingamegui.py]: Over the whole codebase that is.

File: horizons/world/player.py (hunk 5 above)
Comment (MarkusHackspacher) [inline horizons/world/player.py]: This are the class description and so are the right place above class and not above __init__, so I lern. And in this project are the class description above class, too. And no comment above __init__.  So I make this consistent.
By the way, Epydoc is outdated, a alternative is pdoc https://github.com/BurntSushi/pdoc or Sphinx.

File: horizons/world/player.py (hunk 6 above)
Comment (MarkusHackspacher) [inline horizons/world/player.py]: I don´t know.

File: horizons/world/player.py (hunk 4 above)
Comment (MarkusHackspacher) [inline horizons/world/player.py]: I split only `assert isinstance(name, basestring) and name` in two assert.
If you want, I can add.

File: horizons/component/tradepostcomponent.py (hunk 2 above)
Comment (MarkusHackspacher) [inline horizons/component/tradepostcomponent.py]: Yes, I´m sorry.

File: horizons/world/player.py (hunk 4 above)
Comment (squiddy) [inline horizons/world/player.py]: I think we should leave them as they are and discuss in general how to proceed with asserts. I've heard approval of @ChrisOelmueller on IRC to remote isinstance checks already, maybe we should drop type-checks alltogether.

With the move to python3 we *could* make use of [PEP-484](https://www.python.org/dev/peps/pep-0484/) and/or [mypy](http://mypy-lang.org/index.html) or similar.

File: horizons/world/__init__.py (hunk 1 above)
Comment (MarkusHackspacher) [inline horizons/world/__init__.py]: The class will open with None. I have no explanations, just look kind of strange.
```
horizons/world/__init__.py", line 741, in load_raw_world
    world = World(session=None)
```

File: horizons/world/player.py (hunk 5 above)
Comment (MarkusHackspacher) [inline horizons/world/player.py]: > Probably should look at the rest of the codebase to see if it is really consistent.

I try to find it with:
`git grep -A 1 __init__ | grep -B 1 """` and `git grep -A 1 class | grep -B 1 """`
but this command are not working.

File: horizons/world/__init__.py (hunk 1 above)
Comment (squiddy) [inline horizons/world/__init__.py]: It's a test and the session was not needed at this point, so the person didn't bother to pass in a proper Session object.

Which is okay, and demonstrates the point about type-checks hindering testability. :)

Comment (MarkusHackspacher) [inline horizons/world/__init__.py]: But this case happen at normal game play, not only at tests.
Select a new game and click the random map.
 ```
File "/unknown-horizons/horizons/gui/modules/singleplayermenu.py", line 425, in _update_map_preview
    world = load_raw_world(map_file)
 File "/unknown-horizons/horizons/world/__init__.py", line 741, in load_raw_world
    world = World(session=None)
 File "/unknown-horizons/horizons/world/__init__.py", line 86, in __init__
    assert isinstance(session, horizons.session.Session)
AttributeError: 'module' object has no attribute 'session```

File: horizons/world/player.py (hunk 5 above)
Comment (squiddy) [inline horizons/world/player.py]: Your shell is probably complaining about quotes. Make sure to wrap those """ in quotes again, like this:

`git grep -A 1 class | grep _B 1 '"""'` (note the ')

File: horizons/world/__init__.py (hunk 1 above)
Comment (squiddy) [inline horizons/world/__init__.py]: You need to import it first.

Comment (squiddy) [inline horizons/world/__init__.py]: > The class will open with None.

I misread the code, I see it's beeing used in `load_raw_world`, probably for generating the map preview. Other than somehow refactoring that code, just delete the assertion.

File: horizons/world/player.py (hunk 5 above)
Comment (MarkusHackspacher) [inline horizons/world/player.py]: the facts are:
```
$ git grep -A 1 class | grep -B 1 '"""' | wc -l
1636
$ git grep -A 1 _init_ | grep -B 1 '"""' | wc -l
242
```

Comment (squiddy) [inline horizons/world/player.py]: Well, that only checks whether there is any docstring at all in `__init__`. Obviously we have more docstrings in classes.

Oh well, I'm kinda getting lost with the goal of this PR here honestly.

File: horizons/world/player.py (hunk 4 above)
Comment (MasterofJOKers) [inline horizons/world/player.py]: A question about type hints: Isn't that basically the same as doing `isinstance()` all over?

File: horizons/world/player.py (hunk 5 above)
Comment (MasterofJOKers) [inline horizons/world/player.py]: All this tells is, that sometimes `__init__` is not documented. It doesn't say anything about parameters documented directly at the beginning of a class definition.

File: horizons/world/player.py (hunk 4 above)
Comment (squiddy) [inline horizons/world/player.py]: Type hints aren't enforced by the interpreter AFAIK. Third-party libraries may use that to provide some kind of static type checking.

I see it as a way of documentation, and possibly later, some static type checks on the codebase (not during runtime).

Comment (squiddy) [inline horizons/world/player.py]: A way of documentation that is part of python now and doesn't require (to me) ugly markup like Epydoc.

File: horizons/world/player.py (hunk 5 above)
Comment (MarkusHackspacher) [inline horizons/world/player.py]: It is ok, if I make the changes back and I make a separate PR for this?

**Comment by MarkusHackspacher** [comment]
Have change your requests. @MasterofJOKers

**Comment by squiddy** [comment]
This looks okay, with the exception of the now active assertion in player.py, which @MarkusHackspacher doesn't know whether it's safe or not: https://github.com/unknown-horizons/unknown-horizons/pull/2503#discussion_r82518318

Similar code in world/__init__.py was removed entirely, so we should do the same in player.py.

**Comment by MarkusHackspacher** [comment]
This assert is save now.
And it was commented out earlier.
Something in the program must have changed that it works, but I do not know what.
